### PR TITLE
Make generic audio stats and check stats before ffmpeg filters

### PR DESCRIPTION
**Pull request by khumairraj**
Add stats function so that all audio formats can be handled.
Check conditions before processing audio with ffmpeg

### CODE REVIEW

File: hearpreprocess/util/audio.py [hunk 1]
```diff
@@ -3,129 +3,168 @@
...
-            "-c:a",
-            "pcm_f32le",
-            str(out_file),
-        ],
-        stdout=subprocess.DEVNULL,
-        stderr=subprocess.DEVNULL,
-    )
-    # Make sure the return code is 0 and the command was successful.
-    assert ret == 0, f"ret = {ret}"
-
```
Comment (turian) [inline hearpreprocess/util/audio.py:40]: I'm not so sure about changing these ffmpeg commnads. They were given to us by Justin Saloman as using the ffmpeg intereface to get speed but the high-quality sox backends. So unless we match these commands exactly (and you show each flag above in the command below), I vote against this change.

Why did you change this?

File: hearpreprocess/util/audio.py [hunk 2]
```diff
@@ -3,129 +3,168 @@
...
-    # print(" ".join(["ffmpeg", "-y", "-i", in_file,
-    #    "-ac", "1", "-c:a", "pcm_f32le", out_file]))
-    ret = subprocess.call(
-        ["ffmpeg", "-y", "-i", in_file, "-ac", "1", "-c:a", "pcm_f32le", out_file],
-        stdout=devnull,
-        stderr=devnull,
-    )
-    # Make sure the return code is 0 and the command was successful.
-    assert ret == 0
+    # Get the audio stats for the audio file
```
Comment (turian) [inline hearpreprocess/util/audio.py:22]: Same here and throughout the file

File: hearpreprocess/util/audio.py [hunk 3]
```diff
@@ -3,129 +3,168 @@
...
-        {
+        # Percentile duration of the audio
+        **{
             f"{str(p)}th percentile dur(sec)": np.percentile(durations, p)
             for p in [10, 25, 75, 90]
-        }
-    )
-    json.dump(stats, open(out_file, "w"), indent=True)
+        },
+    }
```
Comment (turian) [inline hearpreprocess/util/audio.py:168]: 1) Can you put the label counts in this file too?

2) Can you combine all the files at the end with one final job so there is like one JSON file that contains like all the information about a task preprocessing, so you can quickly look at one file and debug if it's correct?

File: hearpreprocess/util/audio.py (hunk 2 above)
Comment (khumairraj) [reply hearpreprocess/util/audio.py:22]: This function is no more in use so I removed it. We only use the `mono_wav_and_fix_duration`

File: hearpreprocess/util/audio.py (hunk 1 above)
Comment (khumairraj) [reply hearpreprocess/util/audio.py:40]: I broke down the command into pieces so that we can select filters to apply depending on the need. If the audio already has the required `duration` or the audio already is `mono` and `.wav` extension, we will create a symlink, otherwise we will add the corresponding things in the chain for ffmpeg to process
```
[
            "ffmpeg",
            "-y",  --> handled by the overwrite_output()
            "-i", --> handled by the input()
            str(in_file),
            "-filter_complex", --> creates a filter chain, this is done by the python command by chaining the steps with the dot
            f"apad=whole_dur={duration},atrim=end={duration}", --> done by the apad and atrim filter
            "-ac", --> this sets the output to mono audio. done in the output
            "1",
            "-c:a", --> This sets the output codec which is done by the acodec filter
            "pcm_f32le",
            str(out_file),
        ],
```
becomes basically this 
```
ffmpeg.input(in_file).audio
.filter("apad", whole_dur=duration)
.filter("atrim", end=duration)
.output(out_file, f="wav", acodec="pcm_f32le", ac=1)
.overwrite_output()
.run(quiet=True)
```

Comment (khumairraj) [reply hearpreprocess/util/audio.py:40]: The other command which I changed was --

```
[
            "ffmpeg",
            "-y",
            "-i",
            in_file,
            "-af", --> helps to create chained filter which ffmpeg python does
            "aresample=resampler=soxr", --> by the filter aresample
            "-ar", --> in the output
            str(out_sr),
            out_file,
        ],
```
to 
```
  ffmpeg.input(in_file)
  .filter("aresample", resampler="soxr")
  .output(out_file, ar=out_sr)
  .overwrite_output()
  .run(quiet=True)
```

**Review by turian — APPROVED** [review]

File: hearpreprocess/util/audio.py (hunk 3 above)
Comment (turian) [reply hearpreprocess/util/audio.py:168]: For 2 above, we'll do it separately using this https://github.com/neuralaudio/hear-preprocess/issues/24

File: hearpreprocess/util/audio.py [hunk 4]
```diff
@@ -14,17 +14,30 @@
...
-    duration_filter_flag = audio_stats["duration"] != duration
-    mono_filter_flag = not audio_stats["mono"]
-    ext_filter_flag = audio_stats["ext"].lower() != ".wav"
+    # Get the filters to apply to the audio. 
+    # If ffmpeg probe is unable to fetch the audio stats
+    # in which case the stats will be empty
+    # we will make all the filters as True
+    if audio_stats:
+        duration_filter_flag = audio_stats["duration"] != duration
+        mono_filter_flag = not audio_stats["mono"]
```
Comment (turian) [inline hearpreprocess/util/audio.py:36]: _flag is a terrible variable name because it doesn't tell you the orientation.

What about "_incorrect"?

File: hearpreprocess/util/audio.py [hunk 5]
```diff
@@ -14,17 +14,30 @@
 
 def mono_wav_and_fix_duration(in_file: str, out_file: str, duration: float):
```
Comment (turian) [inline hearpreprocess/util/audio.py:15]: Please give a return signature for files here

File: hearpreprocess/util/audio.py [hunk 6]
```diff
@@ -98,12 +112,8 @@ def get_audio_stats(in_file: Union[str, Path]):
...
         }
-    except ffmpeg.Error as e:
-        print(
-            "Skipping audio file for stats calculation. "
-            "Audio path: {file_path}"
-            f"Error: {e}"
-        )
+    except (ffmpeg.Error, KeyError):
+        #Skipping audio file for stats calculation.
         audio_stats = {}
```
Comment (turian) [inline hearpreprocess/util/audio.py:117]: Return None on error

File: hearpreprocess/util/audio.py [hunk 7]
```diff
@@ -3,129 +3,178 @@
...
-import subprocess
-from collections import Counter
+from collections import Counter, defaultdict
 from pathlib import Path
-from typing import List, Optional, Union
+from typing import List, Optional, Union, Dict
 
 import numpy as np
-import soundfile as sf
+import ffmpeg
```
Comment (turian) [inline hearpreprocess/util/audio.py:11]: You have to add this to setup.py and requirements.txt

File: hearpreprocess/util/audio.py [hunk 8]
```diff
@@ -3,129 +3,178 @@
...
+        # Add duration commands if required
+        if duration_filter_flag:
+            cmd_chain = cmd_chain.filter("apad", whole_dur=duration).filter(
+                "atrim", end=duration
+            )
+        # Add mono flag if required
+        if mono_filter_flag:
+            cmd_chain = cmd_chain.output(out_file, f="wav", acodec="pcm_f32le", ac=1)
+        else:
+            cmd_chain = cmd_chain.output(out_file, f="wav", acodec="pcm_f32le")
```
Comment (turian) [inline hearpreprocess/util/audio.py:56]: If you have to specify the output then just specify it is mono regardless and remove the if

File: hearpreprocess/util/audio.py [hunk 9]
```diff
@@ -3,129 +3,178 @@
...
+            _ = cmd_chain.overwrite_output().run(quiet=True)
+        except ffmpeg.Error as e:
+            print(
+                "Please check the console output for ffmpeg to debug the "
+                "error in mono wav and fix duration: ",
+                f"Error: {e}",
+            )
+            raise
+    else:
+        # If the file already has the desired duration and is mono wav, make a symlink
```
Comment (turian) [inline hearpreprocess/util/audio.py:68]: Again. We don't unlink files. We throw an error. This is a bug.

File: setup.py [hunk 10]
```diff
@@ -39,6 +39,7 @@
         "spotty",
         "tqdm",
         "scikit-learn>=0.24.2",
+        "ffmpeg-python",
```
Comment (turian) [inline setup.py:42]: and requirements.txt

**Comment by khumairraj** [comment]
#36 Please refer to this
